roles/confluent.common/tasks/remove_packages.yml

```diff
@@ -0,0 +1,79 @@
+---
+- name: Get Package Facts
+  package_facts:
+
+- name: Determine if Confluent Platform Package Version Will Change
+  set_fact:
+    confluent_package_version_changed: >-
+      {{ ansible_facts.packages['confluent-common'] is defined
+          and ansible_facts.packages['confluent-common'][0].version
+            not in [confluent_package_version, confluent_full_package_version] }}
+
+- debug:
+    var: confluent_package_version_changed
+
+- name: Get installed Confluent packages
+  set_fact:
+    confluent_packages_actual: >
+      {{ ansible_facts.packages | dict2items | map(attribute='key') | select('match', 'confluent-') | list }}
+
+- debug:
+    var: confluent_packages_actual
+
+- name: Determine desired Confluent packages
+  set_fact:
+    confluent_packages_desired: >
+      {{ confluent_packages_desired | default([]) | union(item.value.packages) }}
+  loop: "{{ confluent_components | dict2items }}"
+  loop_control:
+    label: "{{ item.key }}"
+  when: inventory_hostname in groups.get(item.key, [])
+
+- debug:
+    var: confluent_packages_desired
+
+- set_fact:
+    confluent_packages_removed: >
+      {{ (confluent_packages_actual | difference(confluent_packages_desired))
+          if not confluent_package_version_changed|bool
+          else confluent_packages_actual }}
+
+- debug:
+    var: confluent_packages_removed
+
+- name: Get service facts
+  service_facts:
+  when: confluent_packages_removed|length > 0
+
+- debug:
+    var: ansible_facts.services
+  when: confluent_packages_removed|length > 0
+
+- name: Stop Service before Removing Packages
```

Comment:
Im not so sure about stopping other services. At least for this PR. My thought is if someone is running the `kafka_broker.yml` playbook against colocated zk/kafka clusters then it will stop zk and then install kafka and fail bc zk never got restarted. I think there is great work around checking if package versions changed or if there is a difference between packages intended to be installed and actually installed.

But at least for now i think the kafka_broker role should only affect kafka_broker things.

There are probably more factors to think about for the colocated upgrade... One of which is property files get overwritten when you upgrade. You can't just run
```
yum install confluent-server
systemctl start confluent-server
```
bc the server.properties gets overwritten. We should most likely consider the colocated situation in a follow up PR. WDYT?

Reply:
Yes, there seems to be more delicate matters to deal with - that I do not know very well. 😅 What do you think I should do now, then? Find all the packages to remove, even if colocated, or just remove the ones that applies to the role? And just stop the service for the role applied. Or ignore colocated totally for now?

WDYT about using the `service_facts` module in case packages are about to be removed?

Reply:
I'd actually say roll back a few commits before it got this complex and just invoke this task file with 
```
- name: Stop Service and Remove Packages
  include_role:
    name: confluent.common
    tasks_from: remove_packages.yml
  when: installation_method == "package"
  vars:
    service_name: "{{zookeeper_service_name}}"
    package_list: "{{zookeeper_packages}}"
  tags: package
```

And we can follow this logic:
Stop service if:
- versions change
- there will be a confluent-kafka -> confluent-server switch

Remove all packages if:
- version change

Remove confluent-kafka/server package if it should be removed

And thats it. It wont concern colocated, which we can go back to brainstorming about. but at least it will do upgrades/downgrades/confluent-server->confluent-kafka switches on non colocated. thoughts?

Reply:
Ok, I'll try to do my best and mark it ready for review (non-draft) when done. 😄

Reply:
I am uncertain if the current approach is acceptable. I am quite sure it will work good on non colocated, but on colocated (zookeeper + kafka_broker) it is definitely not idempotent. That environment is what I have to test in right now, sorry. 😢 When running colocated, this happens when applying the `zookeeper` role:

```log
TASK [confluent.common : Debug Confluent Packages to Remove] *******************
ok: [h1-a-strm-brd81.statnett.no] => 
  msg:
  - confluent-rebalancer
  - confluent-security
```

These packages are the `kafka_broker` packages that are not part of the intersection of packages with `zookeeper`, so that is expected. The removed packages will be re-installed by the `kafka_broker` role and start/restart it.

I'll have to sleep on it. Please let me know if you have some suggestions apart from those already mentioned....

Reply:
Think I got it now. At least it is idempotent for colocated also.